### PR TITLE
[PUBDEV-4572] Register XGBoost REST API only when backend lib exists

### PR DESCRIPTION
**Pull request by jakubhava**
This solves the issue that Flow didn't show the XGBoost, but it was still reported via registered extensions.

### CODE REVIEW

**Comment by mmalohlava** [comment]
@jakubhava it is good change, however, what do you think about creating Core XGboost extension which will be configured based on availability native lib? There is still question how to make a weak connection between core extension and rest api extension.

**Comment by jakubhava** [comment]
@mmalohlava if I understand it correctly, you mean on the Gradle level? I think that would be the best 👍

**Comment by jakubhava** [comment]
Ups, I understood it in a wrong way. Yup, core extension make sense!

**Comment by jakubhava** [comment]
@mmalohlava It's implemented. There are actually 2 commits. First is implementing the fix and the second is doing the following:

 - Use RestApiContext inside registerEndpoint method -> Less static calls, easier to change backend in the future
 - Create AlgoAbstractRegister with method registering algos. It was previously in the AbstractRegister where it
 semantically does not belong
 - Remove RegisterSourceRoots from extensions. It's not rest api extension. Instead call it explicitly in the  same location where it would be called as the extension

It's easy to separate to another PR.  The changes above use less static calls and unify registering endpoints.

**Comment by jakubhava** [comment]
Rebased on current master

File: h2o-core/src/main/java/water/ExtensionManager.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+
+import water.api.RequestServer;
+import water.api.RestApiExtension;
+import water.api.SchemaServer;
+import water.util.Log;
+
+import java.util.*;
+
+public class ExtensionManager {
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/ExtensionManager.java]: It is better to use singleton approach - look at PersistManager

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostExtension.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  public static String NAME = "XGBoost";
+
+  @Override
+  public String getExtensionName() {
+    return NAME;
+  }
+
+  @Override
+  public boolean isEnabled() {
+//Check if multinode (XGBoost will not work for multinode)
```
Comment (mmalohlava) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostExtension.java]: Formatting.

**Comment by abal5** [comment]
@h2o-ops please test!

**Comment by jakubhava** [comment]
@h2o-ops please test!

**Pull request merged**